### PR TITLE
sql, server: observability replace role options with system privileges

### LINKED ISSUE #83843: Replace relevant role options with equivalently named system privileges
<p data-pm-slice="1 4 []">Create system privileges for some of the existing role options. The relevant role options need to have an equivalent system privilege based on the following table:</p>

Role Option | Inheritable
-- | --
CREATEROLE / NOCREATEROLE | Unchanged
PASSWORD | Unchanged
LOGIN / NOLOGIN | Unchanged
VALIDUNTIL | Unchanged
CONTROLJOB / NOCONTROLJOB | Replace with per job grants Example: GRANT PAUSE ON JOB TYPE BACKUP
CONTROLCHANGEFEED / NOCONTROLCHANGEFEED | Changefeeds are created per table and write to a sink. We can support CHANGEFEED privilege on tables or on the database and also introduce grants on sinks
CREATEDB / NOCREATEDB | Unchanged (Is a Postgres roleoption)
CREATELOGIN / NOCREATELOGIN | Unchanged (?)
VIEWACTIVITY / NOVIEWACTIVITY | Replace with system level grant
CANCELQUERY / NOCANCELQUERY | Replace with system level grant
MODIFYCLUSTERSETTING / NOMODIFYCLUSTERSETTING | Replace with system level grant
VIEWACTIVITYREDACTED / NOVIEWACTIVITYREDACTED | Replace with system level grant
SQLLOGIN / NOSQLLOGIN | Unchanged
VIEWCLUSTERSETTING / NOVIEWCLUSTERSETTING | Replace with system level grant

Jira issue: CRDB-17300

Epic CRDB-14105

### PR DESCRIPTION
**Pull request by Santamaura**
This patch introduces system privileges that exist currently as
role options. For observability they are the following: VIEWACTIVITY,
VIEWACTIVTYREDACTED, VIEWCLUSTERSETTING, CANCELQUERY and NOSQLLOGIN.
Users should now do GRANT SYSTEM `privilege` TO `role`; instead of
using role options.
Brand new system privileges for observability will be introduced in
a subsequent PR.
Any additional gating with these system privileges beyond the basic
ones made in this PR will be done in a subsequent PR.

Resolves https://github.com/cockroachdb/cockroach/issues/83843

Release note (sql change): introduced
VIEWACTIVITY, VIEWACTIVITYREDACTED,
VIEWCLUSTERSETTING, CANCELQUERY,
and NOSQLLOGIN as system privileges.

### CODE REVIEW

**Comment by rafiss** [comment]
drive-by comment -- i feel like this change deserves a release note

**Review by RichardJCai — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84198)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh, @Santamaura, and @zachlite)

___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7S_TTN5Ba8M119Rn0F):*
I'd add a note that you should now do `GRANT SYSTEM ... TO ...` instead of using the role options.
___
*[`pkg/server/admin.go` line 1860 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7S_rMd5hFLNIkeNZfN:-N7S_rMd5hFLNIkeNZfO:b-9nhp8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cad2efd41b90d3dee6846a943e49613a8b520/pkg/server/admin.go#L1860)):*
> ```Go
> 						privilege.VIEWCLUSTERSETTING, privilege.MODIFYCLUSTERSETTING)
> 				}
> 			} else {
> ```

nit: Don't need the else if we're just returning
___
*[`pkg/server/admin.go` line 3474 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7S_pR-36Rd9-cqFLoV:-N7S_pR-36Rd9-cqFLoW:b-am69eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cad2efd41b90d3dee6846a943e49613a8b520/pkg/server/admin.go#L3474)):*
> ```Go
> 						privilege.VIEWACTIVITY)
> 				}
> 			} else {
> ```

nit: Don't need the else if we're just returning

Ditto for other cases below
___
*[`pkg/server/admin.go` line 3642 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7S_vaf02MWVQSo80Zu:-N7S_vaf02MWVQSo80Zv:b-2imm9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cad2efd41b90d3dee6846a943e49613a8b520/pkg/server/admin.go#L3642)):*
> ```Go
> // privilege provided.
> func (c *adminPrivilegeChecker) checkHasSystemPrivilege(
> 	ctx context.Context, user username.SQLUsername, privilege string,
> ```

can `privilege string` become `privilege privilege.Kind`?

I'm kind of okay with this initially but this is not ideal since it'll not use the cache which I'm implementing in a separate PR. Can you add a todo to update this to do the below?

Also can you note what the issue was with creating a planner.

```
desc := syntheticprivilege.GlobalPrivilegeObject.GetPrivilegeDescriptor(ctx, ie)
desc.CheckPrivilege()
...
```

**Review by koorosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84198)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Santamaura and @zachlite)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7V0K-iB4E-nd-bhSWX:-N7V0K-iB4E-nd-bhSWY:b-1kg1bv) (no related file):*
As far as I understand, system privileges should replace existing role options, but current logic requires that user has both assigned role options **and** system privileges. 
I suggest having the following workflow:
- check the system privileges table first (for supported versions) and skip role options validation if user has explicitly defined system privilege; In this case system privileges take precedence over role options.
- if the user doesn't have appropriate privilege **then** fallback to role options validation;

___

File: pkg/server/admin.go [hunk 1]
```diff
@@ -3575,6 +3635,26 @@ func (c *adminPrivilegeChecker) hasRoleOption(
...
+// checkHasSystemPrivilege is a helper function which executes
+// a query to see if rows return non-nil based on the
+// privilege provided.
+func (c *adminPrivilegeChecker) checkHasSystemPrivilege(
+	ctx context.Context, user username.SQLUsername, privilege string,
+) (bool, error) {
+	row, err := c.ie.QueryRowEx(
+		ctx, "check-has-system-privilege", nil, /* txn */
+		sessiondata.InternalExecutorOverride{User: username.RootUserName()},
+		`SELECT * FROM system.privileges WHERE username = $1 AND privileges @> ARRAY[$2]`,
```
Comment (RichardJCai) [inline pkg/server/admin.go:3647]: One more thing, this doesn't account for inheritance.

**Review by Santamaura — COMMENTED** [review]
Added

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-:-N7rB7_4BDnSZ-sz_im_:b-8b11j9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh, @RichardJCai, and @zachlite)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7V0K-iB4E-nd-bhSWX:-N7rByOLDi5HStK_mcg9:b-930p8c) (no related file):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

As far as I understand, system privileges should replace existing role options, but current logic requires that user has both assigned role options **and** system privileges. 
I suggest having the following workflow:
- check the system privileges table first (for supported versions) and skip role options validation if user has explicitly defined system privilege; In this case system privileges take precedence over role options.
- if the user doesn't have appropriate privilege **then** fallback to role options validation;
</blockquote></details>

Moved stuff around so that the logic reflects this idea.

___
___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7S_TTN5Ba8M119Rn0F):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I'd add a note that you should now do `GRANT SYSTEM ... TO ...` instead of using the role options.
</blockquote></details>

Done
___
*[`pkg/server/admin.go` line 3474 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7S_pR-36Rd9-cqFLoV:-N7rDBQO9Wyp2n0ez3m5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cad2efd41b90d3dee6846a943e49613a8b520/pkg/server/admin.go#L3474)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: Don't need the else if we're just returning

Ditto for other cases below
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 3642 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7S_vaf02MWVQSo80Zu:-N7rCkxo2Ti4ywnmBfXt:bp340wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cad2efd41b90d3dee6846a943e49613a8b520/pkg/server/admin.go#L3642)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

can `privilege string` become `privilege privilege.Kind`?

I'm kind of okay with this initially but this is not ideal since it'll not use the cache which I'm implementing in a separate PR. Can you add a todo to update this to do the below?

Also can you note what the issue was with creating a planner.

```
desc := syntheticprivilege.GlobalPrivilegeObject.GetPrivilegeDescriptor(ctx, ie)
desc.CheckPrivilege()
...
```
</blockquote></details>

Per our discussion, switched to using `NewInternalPlanner` wrapped in a function so this is no longer relevant.
___
*[`pkg/server/admin.go` line 3647 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N6itR-JSA_nB0HDvAFl-r2-3647:-N7rD-fK1eHa05SjU0wu:b2nhb2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cad2efd41b90d3dee6846a943e49613a8b520/pkg/server/admin.go#L3647)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

One more thing, this doesn't account for inheritance.
</blockquote></details>

Ack

**Review by koorosh — COMMENTED** [review]
> Reviewed 1 of 12 files at r1, 2 of 5 files at r2, 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84198)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh, @RichardJCai, @Santamaura, and @zachlite)

___
*[`pkg/server/admin.go` line 1838 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7yEncC95GEBJqubnfv:-N7yEncC95GEBJqubnfw:bf6l8f7) ([raw file](https://github.com/cockroachdb/cockroach/blob/35cb78cf4d971a1e362ac46253c54101dcbf605e/pkg/server/admin.go#L1838)):*
> ```Go
> 		if s.st.Version.IsActive(ctx, clusterversion.SystemPrivilegesTable) {
> 			hasView = s.checkHasSystemPrivilege(ctx, user, privilege.VIEWCLUSTERSETTING)
> 			hasModify = s.checkHasSystemPrivilege(ctx, user, privilege.MODIFYCLUSTERSETTING)
> ```

nit. It doesn't relate to this PR specifically, but it might be useful to introduce an extended version of `heckHasSystemPrivilege` to check the list of privileges and return a single result if some of the provided privileges exist (ie `checkHasSomeSystemPrivileges`)
``` go
hasPrivilege := s.checkHasSomeSystemPrivileges(ctx, user, []privilege.Kind{privilege.VIEWCLUSTERSETTING, privilege.MODIFYCLUSTERSETTING})
if hasPrivilege {
  ...
}
```
___
*[`pkg/sql/crdb_internal.go` line 1565 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7yqEYX0KVFW2kQPFWR:-N7yqEYX0KVFW2kQPFWS:bhd3p9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/35cb78cf4d971a1e362ac46253c54101dcbf605e/pkg/sql/crdb_internal.go#L1565)):*
> ```Go
> 			hasView := false
> 			if p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.SystemPrivilegesTable) {
> 				hasModify = p.CheckPrivilege(ctx, syntheticprivilege.GlobalPrivilegeObject, privilege.MODIFYCLUSTERSETTING) == nil
> ```

Ignored possible errors here. `p.CheckPrivilege` can return error that is not handled.

**Review by Santamaura — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84198)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh, @RichardJCai, and @zachlite)

___
*[`pkg/server/admin.go` line 1838 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7yEncC95GEBJqubnfv:-N8-6EkN76be2b1YBvWE:b42nh3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/35cb78cf4d971a1e362ac46253c54101dcbf605e/pkg/server/admin.go#L1838)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

nit. It doesn't relate to this PR specifically, but it might be useful to introduce an extended version of `heckHasSystemPrivilege` to check the list of privileges and return a single result if some of the provided privileges exist (ie `checkHasSomeSystemPrivileges`)
``` go
hasPrivilege := s.checkHasSomeSystemPrivileges(ctx, user, []privilege.Kind{privilege.VIEWCLUSTERSETTING, privilege.MODIFYCLUSTERSETTING})
if hasPrivilege {
  ...
}
```
</blockquote></details>

I could see that being potentially useful but maybe it should be added as needed.
___
*[`pkg/sql/crdb_internal.go` line 1565 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84198#-N7yqEYX0KVFW2kQPFWR:-N8-6QrK5SfIQ_-FMQPC:b-pq160f) ([raw file](https://github.com/cockroachdb/cockroach/blob/35cb78cf4d971a1e362ac46253c54101dcbf605e/pkg/sql/crdb_internal.go#L1565)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Ignored possible errors here. `p.CheckPrivilege` can return error that is not handled.
</blockquote></details>

`p.CheckPrivilege` only returns an error or `nil` so regardless of how the logic is shuffled around we can't return prematurely here with the error since we need to fallback to checking the role options

File: pkg/sql/logictest/testdata/logic_test/system_privileges_mixed [hunk 2]
```diff
@@ -1,8 +1,8 @@
 # LogicTest: local-mixed-22.1-22.2
-
+# test MODIFYCLUSTERSETTING
```
Comment (RichardJCai) [inline pkg/sql/logictest/testdata/logic_test/system_privileges_mixed:2]: nit: remove this line

**Review by RichardJCai — APPROVED** [review]
Nice work!

File: pkg/ccl/logictestccl/testdata/logic_test/crdb_internal_tenant [hunk 3]
```diff
@@ -8,9 +8,11 @@ SELECT count(distinct(node_id)), count(*)  FROM crdb_internal.node_runtime_info
 query IT
 SELECT node_id, name FROM crdb_internal.leases ORDER BY name
 ----
+0  defaultdb
```
[the remark this answers is not in the document]
Comment (Santamaura) [reply pkg/ccl/logictestccl/testdata/logic_test/crdb_internal_tenant:11]: Yeah still not sure why it had to be updated

File: pkg/sql/logictest/testdata/logic_test/system_privileges_mixed (hunk 2 above)
Comment (Santamaura) [reply pkg/sql/logictest/testdata/logic_test/system_privileges_mixed]: Removed

**Comment by Santamaura** [comment]
TFTRs

**Pull request merged**